models/mli/fairness/model_fairxgb.py

```diff
@@ -0,0 +1,506 @@
+
+                X_enc=self.enc.transform(X[self.X_categorical]).toarray()
+
+                X = pd.concat([X[self.X_numeric], pd.DataFrame(X_enc, columns=self.encoded_categories)], axis=1)
+
+        # Replace missing values with a missing value code
+        if len(self.X_numeric) > 0:
+   
+            for colname in self.X_numeric:
+                X[colname] = list(X[colname].fillna(-999)) 
+
+        # Make sure the target that represents a positive outcome is 1
+        if self.positive_target == 0:
+            y = 1 - y
+        X_full = X.copy()
+        y_full = y.copy()
+        
+        # Set up a validation step to find the optimal number of trees
+        X_valid = X.iloc[int(0.7*len(X_full)): , :]
+        y_valid = y[int(0.7*len(X_full)):]
+        X = X.iloc[0:int(0.7*len(X_full)), :]
+        y = y[0:int(0.7*len(X_full))]
+
+        if self.protected != "none":
+            # Set the protected group to 0 and all others 1          
+            protected_full = [int(item) for item in ~(np.array(X_full[self.protected]) == self.protected_label)]
+            protected_train = [int(item) for item in ~(np.array(X[self.protected]) == self.protected_label)]
+            protected_valid = [int(item) for item in ~(np.array(X_valid[self.protected]) == self.protected_label)]
+        else:
+            mu=0
+            protected_full = []
+            protected_train =  []
+            protected_valid =  []
+            
+            
+        # Remove the protected value from the model
+        if self.protected != "none":
+                      
+            X = X.drop(self.protected, axis=1)
+        
+            X_full = X_full.drop(self.protected, axis=1)
+        
+            X_valid= X_valid.drop(self.protected, axis=1)
+
+        d_train = xgb.DMatrix(X, label=y, missing=np.nan)
+    
+        d_valid = xgb.DMatrix(X_valid, label=y_valid, missing=np.nan)
+
+        # Initial run to find the optimal number of trees
+        num_iterations=10000
+        watchlist = [(d_train, 'train'), (d_valid, 'valid')]
+
+        clf = xgb.train(params, d_train, num_iterations, watchlist, feval=fair_metric, verbose_eval=10, obj=fair, early_stopping_rounds=10)
+
+
+
+        # Second xgboost run with the full dataset and optimal number of trees
+        attribute_dict = clf.attributes()
+        new_iterations = int(attribute_dict['best_iteration'])
+
+        d_train = xgb.DMatrix(X_full, label=y_full, missing=np.nan)
+        watchlist = [(d_train, 'train')]
+        clf = xgb.train(params, d_train, new_iterations, watchlist, feval=fair_metric, verbose_eval=10, obj=fair)
+  
+    
+        # Calculate feature importances
+        importances_dict = clf.get_score(importance_type='gain')
+
+        # Make sure the protected group has high feature importance
+        # so that it doesn't get dropped by driverless
+        if self.protected != "none": 
```

Comment:
If this column is truly important, then wouldn't it come up in the feature importance?

Reply:
This is a debiasing recipe, so you wouldn't want to use the protected column in the model.  It isn't used in the model, so it would tend to be removed by driverless.  I'm giving it a high feature importance, so that driverless won't drop it from experiments.  It's not used in the model but is needed for debiasing.

Reply:
Yes, I understand not to use it. I was just trying to figure out the feature importance aspect to this. I am still wondering if it is really needed. How do they do this in the paper? Same way?

Reply:
I've been setting the feature importance to a high value so that driverless will not decide the protected value is unimportant and drop the protected column. That's not an issue in the paper since they are not using driverless.